scipy/optimize/_differentialevolution.py

```diff
@@ -823,6 +957,49 @@ def __del__(self):
         self._mapwrapper.close()
         self._mapwrapper.terminate()
 
+    def _accept_trial(self, energy_trial, feasible_trial, cv_trial,
+                      energy_orig, feasible_orig, cv_orig):
+        """
+        Trial is accepted if:
+        * it satisfies all constraints and provides a lower or equal objective
+          function value, while both the compared solutions are feasible
+        - or -
+        * it is feasible while Xi, G is infeasible,
+        - or -
+        * it is infeasible, but provides a lower or equal constraint violation
+          for all constraint functions.
+
+        Parameters
+        ----------
+        energy_trial : float
+            Energy of the trial solution
+        feasible_trial : float
+            Feasibility of trial solution
+        cv_trial : array-like
+            Excess constraint violation for the trial solution
+        energy_orig : float
+            Energy of the original solution
+        feasible_orig : float
+            Feasibility of original solution
+        cv_orig : array-like
+            Excess constraint violation for the original solution
+
+        Returns
+        -------
+        accepted : bool
+
+        """
+        if feasible_orig and feasible_trial:
+            return energy_trial < energy_orig
+        elif feasible_trial and not feasible_orig:
+            return True
+        elif not feasible_trial and (cv_trial <= cv_orig).all():
```

Comment:
Even though constraints can be `<=` or `>=`, this is expressed the same as in Equation 9. This is valid because of the way the constraint violations `cv` are defined. (If the constraint is violated, `cv` = the magnitude of the constraint violation, otherwise it is zero.) Looks right.

Note that Lampinen is not 100% self-consistent here. At the bottom of page 3 "Then the candidate vector can be considered less infeasible, and thus better than the current vector, if it it does not violate any of the constraints more than the current vector _and if it violate[s] less a least one of the constraints_." That last bit is not reflected in Equation 9 or the English translation below.

Reply:
Not sure what to do if the paper isn't consistent. If one considers that a trial is accepted if `energy_trial <= energy_orig`, then perhaps *"and if it violate[s] less a least one of the constraints"* can be disregarded for the time being?

Reply:
I think so. I think the simple <= here fits with the philosophy of <= for the objective function anyway.